compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp

```diff
@@ -810,11 +810,31 @@ setReductionVectorDistributionConfig(IREE::GPU::TargetAttr target,
       threadLoads /= 2;
     }
   }
+
+  std::optional<int64_t> parallelSize = 1;
+  for (int64_t dim : parallelDims) {
+    if (ShapedType::isDynamic(bounds[dim])) {
+      parallelSize = std::nullopt;
+      break;
+    }
+    *parallelSize *= bounds[dim];
+  }
+
   // Deduce the workgroup size we should use for reduction. Currently a
   // workgroup processes all elements in reduction dimensions. Need to make sure
   // the workgroup size we use can divide the total reduction size, and it's
   // also within hardware limitations.
-  const int64_t maxWorkgroupSize = 1024;
+
+  ArrayRef<int32_t> maxWgSizes = wgp.getMaxWorkgroupSizes();
+  int32_t maxWorkgroupSize = *llvm::max_element(maxWgSizes);
+  const int numComputeUnits = getGPUNumComputeUnits(target).value_or(256);
+  const int numSIMDs = wgp.getSimdsPerWgp().value_or(4);
+
+  // If there is more than enough work to saturate all CUs, use single subgroup
+  // per workgroup
+  if (parallelSize && *parallelSize > numComputeUnits * numSIMDs)
+    maxWorkgroupSize = target.getPreferredSubgroupSize();
```

Comment:
In the future, we should also decide the local splik k factor based on the number of SIMDs. For example, if N=512 and we only have 256 SIMDs, we can split the K dim by two. Maybe leave a TODO for this?

Also nit: the IREE coding style requires braces around all `if`s/`for`s